### PR TITLE
(655a) Select programme for adding accredited programme history

### PR DESCRIPTION
**Pull request by jsrobertson**
## Context
https://trello.com/c/i2gzZOoW/655-add-page-for-selecting-course-name-ui

## Changes in this PR

- Adds the form page for selecting a course to add a programme history to a referral. 
- Adds new `ProgrammeHistoryController` with `new` method to render the form page.

A follow on PR will add `create` method to create to `POST` to `/course-participations` and the functionality to submit the form (with validation).

## Screenshots of UI changes
localhost_3000_referrals_c11fab18-dc8d-420c-9c82-d0edd373732d_programme-history_new

## Release checklist

[Release process documentation](../doc/how-to/perform-a-release.md)

As part of our continuous deployment strategy we must ensure that this work is
ready to be released once merged.

### Pre-merge

- [ ] There are changes required to the Accredited Programmes API for this change to work...
  - [ ] ... and they have been released to production already

### Post-merge

- [x] This adds, extends or meaningfully modifies a feature...
  - [ ] ... and I have written or updated an end-to-end test for the happy path in the [Accredited Programmes E2E repo](https://github.com/ministryofjustice/hmpps-accredited-programmes-e2e)
- [ ] This makes new expectations of the API...
  - [ ] ... and I have notified the API developer(s) of changes to the contract tests (Pact), or the API is already compliant
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-preprod-environment) release to preprod
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-production-environment) release to prod

### CODE REVIEW

File: server/views/referrals/programmeHistory/new.njk [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+            legend: {
+              text: pageHeading,
+              isPageHeading: true,
+              classes: "govuk-fieldset__legend--l"
+            }
+          },
+          hint: {
+            text: "Select the programme"
+          },
+          items: courseOptions.concat({
```
Comment (jsrobertson) [inline server/views/referrals/programmeHistory/new.njk:43]: Will need an integration test

File: server/views/referrals/programmeHistory/new.njk [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  {{ govukInput({
+      id: "otherCourseName",
+      name: "otherCourseName",
+      type: "text",
+      classes: "govuk-!-width-one-third",
+      label: {
+        text: "Enter the programme"
+      }
+    }) }}
+  {% endset -%}
```
Comment (yndajas) [inline server/views/referrals/programmeHistory/new.njk:26]: `-%}` -> `%}`

File: server/views/referrals/programmeHistory/new.njk [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-two-thirds">
+      <form action="#" method="post">
+        {{ govukRadios({
+          name: "programme",
```
Comment (yndajas) [inline server/views/referrals/programmeHistory/new.njk:32]: `courseId`? Or `course` if this isn't the actual field

File: server/views/referrals/programmeHistory/new.njk (hunk 1 above)
Comment (yndajas) [inline server/views/referrals/programmeHistory/new.njk:43]: `courseRadioOptions`?

File: server/controllers/refer/programmeHistoryController.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import type { Request, Response, TypedRequestHandler } from 'express'
+
+import type { CourseService } from '../../services'
+import { CourseUtils, TypeUtils } from '../../utils'
+
+export default class ProgrammeHistoryController {
```
Comment (yndajas) [inline server/controllers/refer/programmeHistoryController.ts:6]: Per previous comment, I think we could be calling this `CourseParticipationsController` per the entity name

File: server/controllers/refer/programmeHistoryController.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  constructor(private readonly courseService: CourseService) {}
+
+  new(): TypedRequestHandler<Request, Response> {
+    return async (req, res) => {
+      TypeUtils.assertHasUser(req)
+
+      const courses = await this.courseService.getCourses(req.user.token)
+
+      res.render('referrals/programmeHistory/new', {
+        courseOptions: CourseUtils.courseRadioOptions(courses),
```
Comment (yndajas) [inline server/controllers/refer/programmeHistoryController.ts:16]: Per previous comment, perhaps `courseRadioOptions`?

File: server/controllers/refer/programmeHistoryController.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+        pageHeading: 'Add Accredited Programme History',
```
Comment (yndajas) [inline server/controllers/refer/programmeHistoryController.ts:17]: `History` -> `history`

File: server/paths/refer.ts [hunk 7]
```diff
@@ -13,6 +13,8 @@ const personPath = peoplePathBase.path(':prisonNumber')
 const referralsPath = path('/referrals')
 const showReferralPath = referralsPath.path(':referralId')
 const referralPersonPath = showReferralPath.path('person')
+const programmeHistoryBasePath = showReferralPath.path('programme-history')
+const newProgrammeHistoryPath = programmeHistoryBasePath.path('new')
```
Comment (yndajas) [inline server/paths/refer.ts:17]: I'm a little less sure about the course participations vs programme history question here - I think `/programme-history` is clearer for the user, but `/programme-history/new` feels a little strange when the entity we're creating is a course participation. I've gone for `/course-participations` in the branch I'm currently working on, but I'm curious what you and @Gweaton think on this one!

Comment (yndajas) [reply server/paths/refer.ts:17]: Let's go with `/programme-history/new` (and `/programme-history`) - just run this by George in an unrelated chat!

File: server/views/referrals/programmeHistory/new.njk [hunk 8]
```diff
@@ -0,0 +1,58 @@
+{% from "govuk/components/back-link/macro.njk" import govukBackLink %}
```
[the remark this answers is not in the document]
Comment (jsrobertson) [reply server/views/referrals/programmeHistory/new.njk:1]: That makes sense 👍

**Review by yndajas — APPROVED** [review]

**Pull request merged**